Check in algo generation code

c1728p9 opened the pull request:
Check in scripts which are able to generate flash algos for supported
targets.

To initially download all packs the following command should be run:
```
>> python extract.py --rebuild_all
```

After that all supported targets can be rebuilt by running:
```
>> python extracts.py
```

Finally, to rebuild an individual target you can used its pack name:
```
>> python extract.py --target STM32F302R8
```

c1728p9 commented:
CC: @0xc0170

theotherjimmy commented on tools/flash_algo/flash_algo.py:
```diff
@@ -0,0 +1,359 @@
+        )
+
+        ro_rw_zi = _find_sections(self.elf, sections_to_find)
+        ro_rw_zi = _algo_fill_zi_if_missing(ro_rw_zi)
```
immediately re-assigning a variable can be confusing. maybe
```python
ro_rw_zi = _algo_fill_zi_if_missing(_find_sections(self.elf, sections_to_find))
```
This?

theotherjimmy commented on tools/flash_algo/extract_targets.py:
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
```
maybe `extract_all.py`?

c1728p9 commented on tools/flash_algo/flash_algo.py:
```diff
@@ -0,0 +1,359 @@
+#!/usr/bin/env python
```
Discussed in person. Summary - long term this file should be moved into arm-pack-manager or another common repo. Right now the master copy lives in the [FlashAlgo](https://github.com/mbedmicro/FlashAlgo/blob/master/scripts/flash_algo.py) repository and this is a copy of it.

c1728p9 commented:
/morph test

0xc0170 commented:
I tested this PR. I rebuilt all devices, and run:

```
python extract.py --target STM32F302R8
python extract.py --target NUCLEO_F401RE
python extract.py --target K64F
```

The last two failed with (same error):

```
tools\arm_pack_manager\__init__.py", line 263, in get_flash_algorthim_binary
    device = self.index[device_name]
KeyError: 'NUCLEO_F401RE'
```

As I understand, this script accepts the name that comes from the pack. I thought it would translate it targets.json target name to the cmsis pack name (as targets contain ``device_name``), if there's none, its not supported. if there is, use that name to find it in the packs.

``STM32F302R8`` worked, but generates two files, whats the difference? I looked at both, the flash start address is different (some numbers differ), but I don't know which one I should use? two  devices with various sectors.

```
 ls output\
STM32F302R8_0.c  STM32F302R8_1.c
```

c1728p9 commented:
The name of the pack target should be used with `--target`, not the name of the mbed device. The easy way to run this script is to run `python extract.py` which will build all supported mbed targets.

The multiple flash algos for the same target are for different regions. The flash_start and flash_size will be different for each of these. Typically one algo will be flash and the others will be regions for configurable settings.

To simplify this I updated the script to only create an algo for the flash region if it can determine this. I also added the option `--all` which will cause all flash algos for a device to be created like before.

c1728p9 commented:
/morph test

0xc0170 commented:
Thanks for the explanation. I pulled the latest update, works for me

LGTM

0xc0170 commented:
I fixed the template plus implementations - to protect with ``DEVICE_FLASH``

/morph test

c1728p9 commented:
Thanks for the update @0xc0170, the changes look good to me.
